scala/scala.bzl

```diff
@@ -384,3 +424,193 @@ def scala_repositories():
+  ctx.action(
+    inputs = ctx.files.srcs,
+    outputs = [ctx.outputs.libarchive],
+    # tar -c src/thrift/com/example/thrift/AStruct.thrift > scrooge_thrifts.scroogelib.tar
+    command = "echo yeah yeah yeah", #TODO do something real :P
+  )
+  transitive_archive_files = []
+  return struct(
+    srcs=ctx.files.srcs,
+    _transitive_archive_files=transitive_archive_files,
+  )
+
+def _gen_scrooge_srcjar_impl(ctx):
+  #TODO would there be a case in which we want/need to allow more jars here?
+  transitive_cjars = [ctx.files._scrooge_core, ctx.files._libthrift, ctx.files._finagle_thrift, ctx.files._scalalib]
+  transitive_cjars = [file for files in transitive_cjars for file in files]
+
+  cjars = ctx.files._scrooge_generator + \
+          ctx.files._scopt + \
+          ctx.files._twitter_util + \
+          ctx.files._util_logging + \
+          ctx.files._scala_parser_combinators + \
+          transitive_cjars
+  _write_manifest(ctx) #TODO do we need something like this?
+
+  #TODO we need to use the gen-file-map for caching
+  #TODO what imports do we need?
+  #TODO do we want an ijar for this as well?
+  cmd = """
+rm -rf {out}_tmp
+set -e
+{java} -classpath "{jars}" com.twitter.scrooge.Main -d {out}_tmp -s -l scala $@
+find {out}_tmp -exec touch -t 198001010000 {{}} \;
+touch -t 198001010000 {manifest}
+{jar} cmf {manifest} {out} -C {out}_tmp .
+rm -rf {out}_tmp
+""".format(
+    java=ctx.file._java.path,
+    out=ctx.outputs.srcjar.path,
+    manifest=ctx.outputs.manifest.path,
+    jar=ctx.file._jar.path,
+    jars=":".join([f.path for f in cjars]),
+  )
+  ctx.action(
+    #TODO we should curate this a bit more deliberately...make sure it is accurate when it all is done
+    inputs=ctx.attr.scrooge_library.srcs +
+        cjars +
+        ctx.files._jdk +
+        ctx.files._scalasdk +
+        [ctx.outputs.manifest, ctx.file._jar],
+    outputs=[ctx.outputs.srcjar], #TODO ijar?
+    command=cmd,
+    progress_message="scrooge %s" % ctx.label,
+    # Since we have access to the graph, we don't have to muck with the
+    # scrooge_library tar, we can just use the sources directly
+    #TODO this will not give us the relative imports, though... we need
+    # to set a root dir for that to work properly
+    arguments=[f.path for f in ctx.attr.scrooge_library.srcs],
+  )
+
+  #TODO need to dive into what is actually necessary where
+  scalaattr = struct(outputs = ctx.outputs.srcjar, #TODO is this good to have here?
+                     transitive_runtime_deps = [], #TODO what are scrooge's runtime deps?
+                     transitive_compile_exports = transitive_cjars, #TODO are all necessary?
+                     transitive_runtime_exports = []
+                     )
+  return struct(scala = scalaattr, srcjar = ctx.outputs.srcjar)
+
+  #TODO the struct that we return from this is where we can set the scala attributes it wants, like the transitive dependencies
+
+
+scrooge_library = rule(
```

Comment:
So I'm following the pattern used here
https://github.com/wt/bazel_thrift/blob/master/tools/thrift.bzl
as well as in genproto, but I think the idea is you have a "library" rule which says "these thrifts form a library," and then specific language library rules which do the code gen. Scrooge library makes the thrifts available as thrifts (and doesn't force the user to muck with issues around dependencies, scala versions, etc), and then it is up to the specific code gen targets to use that information to generate code for a given platform.